### ISSUE TITLE
configpm: POD formatting problems and more

### ISSUE THREAD

**Issue by jkeenan** [Perl/perl5#21045]
The documentation for `configpm` has a number of POD formatting problems.

First, if I run it through the `podchecker` utility shipped with `perl`, I get this output:
```
$ podchecker configpm
*** ERROR: '=item' outside of any '=over' at line 1108 in file configpm
*** ERROR: You forgot a '=back' before '=head1' at line 1145 in file configpm
*** ERROR: unresolved internal link 'taint_disabled' at line 1112 in file configpm
configpm has 3 pod syntax errors.
```
Similarly, if I run this file through the `pod2text` utility, I get this output:
```
$ pod2text < configpm > /tmp/configpm.txt
IO::File=IO(0x5563352d28f0) around line 1108: '=item' outside of any '=over'
IO::File=IO(0x5563352d28f0) around line 1145: You forgot a '=back' before '=head1'
POD document had syntax errors at /home/jkeenan/perl5/perlbrew/perls/perl-5.36.0/bin/pod2text line 75.
```
Second, it appears that several `heredoc` markers "leak" into the generated output.  From the plain-text file generated above:
```
    "taint_support"
        From define: "SILENT_NO_TAINT_SUPPORT" or "NO_TAINT_SUPPORT"

        If this perl is compiled with support for taint mode this variable
        will be set to 'define', if it is not it will be set to the empty
        string. Either of the above defines will result in it being empty.
        This property was added in version 5.37.11. See also
        "taint_disabled".

        EOF_TEXT

        $glossary{t}{taint_disabled} //= <<EOF_TEXT; =item "taint_disabled"

```
... and ...
```
        being a true value. This property was added in 5.37.11. See also
        "taint_support".

        EOF_TEXT

        if ($Opts{glossary}) { foreach my $fc (sort keys %glossary) { print
        CONFIG_POD "=head2 $fc\n\n=over 4\n\n"; foreach my $item (sort keys
        %{$glossary{$fc}}) { print CONFIG_POD $glossary{$fc}{$item}; } print
        CONFIG_POD "=back\n\n"; } }

        print CONFIG_POD <<'ENDOFTAIL';
```
Since this is newly added documentation, I'm not exactly sure what the intended output should be.

Third, the POD formatting errors are not being detected by any of the tests run via `make test_porting`.

Fourth (and this is somewhat separate from the foregoing), I only stumbled on the formatting problems because I was searching for documentation on `taint_disabled` (and doing that only because of a pull request filed against a CPAN distribution for which I am caretaker).  I would have expected to find such documentation in some file underneath `pod/`, but the only place I could find it was in the generated file `pod/perltoc.pod`.
```
$ ack -l 'taint_(support|disabled)' ./pod
pod/perltoc.pod
```
Even if `configpm` had its POD correctly formatted, I doubt that I would have thought to look there for documentation on these two properties.

**Comment by haarg** [comment]
The pod in `configpm` is not meant to be read directly. It is used to generate `lib/Config.pm`, which won't have the same problems.

It would be better if it was modified to not have anything recognized as pod formatting.  One option for that would be to indent all of the template content, then remove the indentation when generating its output.

**Comment by jkeenan** [comment]
> The pod in `configpm` is not meant to be read directly. It is used to generate `lib/Config.pm`, which won't have the same problems.
> 
Thanks for that clarification. I can appreciate that some of the documentation in a built and/or installed `perl` is generated during configuration and build and therefore might not appear in the core distro in user-friendly form.  That's fine for, say, `pod/perltoc.pod`.

However, a clean checkout is exactly the place where you want information about configuration options like `NO_TAINT_SUPPORT` to be readily accessible.

> It would be better if it was modified to not have anything recognized as pod formatting. One option for that would be to indent all of the template content, then remove the indentation when generating its output.

**Comment by demerphq** [comment]
@jkeenan said:
> The documentation for configpm has a number of POD formatting problems.

As @haarg said, this file is not meant to be viewed directly. It contains code that is used to assemble `Config.pm` from different sources, including the code in [configpm](https://github.com/Perl/perl5/blob/blead/configpm). 

Part of the issue here is that there are ownership issues between some of the data that ends up in `Config.pm` between [./Configure](https://github.com/Perl/perl5/blob/blead/Configure) and the code that lives in the [metaconfig](https://github.com/Perl/metaconfig) repo, and the final set of data that gets injected into `Config.pm`.  Much of our logic assumes that all configuration for a build comes from [./Configure](https://github.com/Perl/perl5/blob/blead/Configure), even though in practice this is not and has never been true as we have long supported `-Accflags` as being an alternative to using Configure prompts to control a compile time flag or setting .

There are a handful of settings that have been introduced by various developers that control how perl is built that are not dealt with at all by [./Configure](https://github.com/Perl/perl5/blob/blead/Configure), or dealt with as an afterthought. The defines that control the taint support that is built into a perl are a good example, as are the various defines relating to hash seeding and function as well as the defines that control various optional behavior in the regular expression engine and maybe other subsystems. Also relevant are defines that are "conditional", and which are defined by logic in our .h files based on other defines.   In some cases there is logic in [./perl.h](https://github.com/Perl/perl5/blob/blead/perl.h) and [./perl.c](https://github.com/Perl/perl5/blob/blead/perl.c) to "surface" these defines into our `perl -V` output, and via special introspection routines that are used by `lib/Config.pm`, however typically there is no support for moving these settings into the `%Config` hash as is expected for most settings.

This is where things get awkward. Most of the data in `Config.pm` and in the `%Config` hash /does/ come from [./Configure](https://github.com/Perl/perl5/blob/blead/Configure) and its probes and Q&A logic from the [./Configure](https://github.com/Perl/perl5/blob/blead/Configure) build step, which is then populated into the file `config.sh`. This is then merged together with documentation and data from [Porting/Glossary](https://github.com/Perl/perl5/blob/blead/Porting/Glossary) and `myconfig.SH`.  [Porting/Glossary](https://github.com/Perl/perl5/blob/blead/Porting/Glossary) is "owned" by @Tux and the [metaconfig](https://github.com/Perl/metaconfig) repo as can be seen by its header:

```
!!!!!!!   DO NOT EDIT THIS FILE   !!!!!!!
This file is built by metaconfig.

This file contains a description of all the shell variables whose value is
determined by the Configure script.  Variables intended for use in C
programs (e.g. I_UNISTD) are already described in config_h.SH.  [`configpm'
generates pod documentation for Config.pm from this file--please try to keep
the formatting regular.]
```

So this means that for a dev to add a new special build mode they have to add code to the perl repo to handle a define, and then they have to go and work together with @tux to get support for it added into [./Configure](https://github.com/Perl/perl5/blob/blead/Configure) in totally different repo. In practice this doesn't happen. @tux is a busy man, and most devs do not bother with updating the [metaconfig](https://github.com/Perl/metaconfig) repo, (the instructions for modifying the [metaconfig](https://github.com/Perl/metaconfig) repository are very intimidating so it is no wonder). Instead they make do with the `-Accflags=-DNEW_SETTING` functionality, and if they are proactive they might update [./perl.h](https://github.com/Perl/perl5/blob/blead/perl.h) or [./perl.c](https://github.com/Perl/perl5/blob/blead/perl.c) so that their new define shows up in `perl -V`, but it is unusual for a dev doing this to actually get the integration with [./Configure](https://github.com/Perl/perl5/blob/blead/Configure) sorted out. So the new setting goes missing from `%Config`  and [Porting/Glossary](https://github.com/Perl/perl5/blob/blead/Porting/Glossary) and is instead awkwardly exposed via `Internals::V()` aka `Config::bincompat_options()` and `Config::non_bincompat_options()`. 

This is precisely what happened with the taint support defines. @tsee added these to core as `NO_TAINT_SUPPORT` and `SILENT_NO_TAINT_SUPPORT`. The way to activate this build mode was to add `-Accflags=SILENT_NO_TAINT_SUPPORT` to the initial [./Configure](https://github.com/Perl/perl5/blob/blead/Configure) command, and this would ensure that the C compiler was provided with these defines as required. 

Later on @neilb did something that is only rarely done, he worked with @tux to add support to the [./Configure](https://github.com/Perl/perl5/blob/blead/Configure) Q&A process to let people enable taint support via a prompt in the [./Configure](https://github.com/Perl/perl5/blob/blead/Configure) stage. This created the `taint_enabled` `%Config` value. However, this work turned out to be problematic in the sense that the implementation essentially assumed that [./Configure](https://github.com/Perl/perl5/blob/blead/Configure) would take over control and ownership of this setting via the new symbol `taint_enabled` in [./config.sh](https://github.com/Perl/perl5/blob/blead/config.sh), without accounting for the fact that adding `-Accflags=SILENT_NO_TAINT_SUPPORT` to [./Configure](https://github.com/Perl/perl5/blob/blead/Configure) on a prompt line was equivalent to answering "yes" to the appropriate prompt in [./Configure](https://github.com/Perl/perl5/blob/blead/Configure). Thus with that patch the 'taint_enabled' setting in `%Config` could end up being "yes" when taint support had actually been disabled, and vice versa. Related to this was that the new setting had a reversed sense from the original. Instead of adding `taint_disabled=yes` and `taint_disabled=silent` as possible configuration values, which would have been more logical, there was only `taint_enabled=yes`. This meant there was no way to tell the difference between "silent" and "noisy" mode of taint removal, which is a very important distinction as passing `-T` to a perl build with `taint_disabled=yes` is an untrappable exception, whereas under `taint_disabled=silent` the `-T` argument would be silently ignored. It also meant that testing if your perl supported taint or not was made more awkward in that you would have to distinguish between different types of false for the setting 'taint_enabled' such that "if it is missing and false it is actually true, but if it is present and false it is false". which is awkward and counter intuitive, the setting should have been called 'taint_disabled' instead, which has much more natural semantics AND allows the user to distinguish between silent mode and non-silent mode via the value. Because of these problems the patch to add support for taint support was reverted and we went back to the old mode of supporting configuring these build options via `-Accflags`.

In 5.37.10 the subject of being able to use `%Config` to identify perls which have taint disabled, and whether that disablement is silent or not came up again in context of work that @DrHyde was doing. In order to detach this from changes to [./Configure](https://github.com/Perl/perl5/blob/blead/Configure) and ensure that the state of this setting was always correct I moved the logic for setting up the `$Config{taint_disabled}` and `$Config{taint_enabled}` to `Config.pm` based on the values returned by `Internals::V()` (which implements `Config::bincompat_options()` and `Config::nonbincompat_options()`) when Config.pm is loaded (similar to how a few other derived settings are handled).  `Internals::V()` returns the *actual* values of the defines as seen by the C compiler itself, not the values that [./Configure](https://github.com/Perl/perl5/blob/blead/Configure) thinks it will see, so it is a much safer mechanism to determine these things anyway, as can be seen by the disagreement bug mentioned above.

But of course, since this is being set up /without/ involving [metaconfig](https://github.com/Perl/metaconfig), and because [metaconfig](https://github.com/Perl/metaconfig) didnt know about "taint_disabled" anyway, and because [metaconfig](https://github.com/Perl/metaconfig) owns [Porting/Glossary](https://github.com/Perl/perl5/blob/blead/Porting/Glossary) there was nowhere to put the documentation for this new settings *but* in [configpm](https://github.com/Perl/perl5/blob/blead/configpm), which is responsible for converting the contents of [Porting/Glossary](https://github.com/Perl/perl5/blob/blead/Porting/Glossary) to the pod contained in lib/Config.pm anyway. When you ran the pod tools on [configpm](https://github.com/Perl/perl5/blob/blead/configpm) you were doing something wrong, that file is not meant to be processed by a pod checker anymore more than [Porting/Glossary](https://github.com/Perl/perl5/blob/blead/Porting/Glossary) is, instead you should have been looking at Config.pm itself, but of course that is a generated file which is not visible to 'git grep'.

Ultimately the problem here is that our approach to [./Configure](https://github.com/Perl/perl5/blob/blead/Configure) and Config.pm and [Porting/Glossary](https://github.com/Perl/perl5/blob/blead/Porting/Glossary) does not reflect our actual development practice for a nontrivial part of our configuration. For many configuration values we use it is critical that [./Configure](https://github.com/Perl/perl5/blob/blead/Configure) know about them and control them, for instance anything that requires compiling a C snippet to test how things work, and it is natural that we let [./Configure](https://github.com/Perl/perl5/blob/blead/Configure) own them from top to bottom. But for feature flag type settings like TAINT_DISABLED this is not the case. For the TAINT_DISABLED flag to be useful [./Configure](https://github.com/Perl/perl5/blob/blead/Configure) need know nothing at all about the new define, it just needs to expose functionality like `-Accflags`.  But because so much of our infra assumes that [./Configure](https://github.com/Perl/perl5/blob/blead/Configure) owns it all there is nowhere to document this new define, to specify what its `%Config` equivalent name is, etc. There is no Porting/FeatureDefine where we can document these things, etc. We have no way to add simple prompting logic to the perl repo to add new Q&A prompts, etc. Instead to add any of this we need to work together with @Tux who is a busy man, and get modifications made to a totally separate repository. As such we have a handful of these feature flag type settings which are under documented, no included in `%Config`, and not really very cleanly handled. IMO it would be MUCH better if we just accept that some of our config needs to be solely owned by the perl.git repo, and have a way to have a sane merger of the bulk which is owned by the [metaconfig](https://github.com/Perl/metaconfig) repo, and the small but important minority that is contained with the perl repo itself. Pretending that all things that are relevant to Config.pm are owned by [./Configure](https://github.com/Perl/perl5/blob/blead/Configure) is a mistake. It just means that it is a much more awkward to add one of these new settings and that it goes only partially supporting by our configuration introspection and documentation and test facilities.

@jkeenan said:
> Third, the POD formatting errors are not being detected by any of the tests run via make test_porting

Well this is not actually the case. `make test_porting` does scan lib/Config.pm for pod errrors. It does not inspect [configpm](https://github.com/Perl/perl5/blob/blead/configpm) because it is not a file which is expected to contain well formed pod, instead it assembles the POD into `lib/Config.pm` from its contents. It only matters that the final file is well formed POD.

@jkeenan said:
> Fourth (and this is somewhat separate from the foregoing), I only stumbled on the formatting problems because I was searching for documentation on taint_disabled (and doing that only because of a pull request filed against a CPAN distribution for which I am caretaker). I would have expected to find such documentation in some file underneath pod/, but the only place I could find it was in the generated file pod/perltoc.pod

I am afraid your assumption is just wrong, *many* of our Config related settings are not documented under pod/. The file you should have been checking is `lib/Config.pod`

@haarg said:
> It would be better if it was modified to not have anything recognized as pod formatting. One option for that would be to indent all of the template content, then remove the indentation when generating its output.

Yeah, I can take care of that. 

@jkeenan said:
> However, a clean checkout is exactly the place where you want information about configuration options like NO_TAINT_SUPPORT to be readily accessible.

Well, now you are moving the goalposts a bit. :-) Your original question was about 'taint_disabled'. However now you are asking about `NO_TAINT_SUPPORT`. The latter define *is* documented in two files in the pod/ directory: [pod/perl5180delta.pod](https://github.com/Perl/perl5/blob/blead/pod/perl5180delta.pod) and [pod/perl53711delta.pod](https://github.com/Perl/perl5/blob/blead/pod/perl53711delta.pod). 

Interestingly it is not mentioned in [INSTALL](https://github.com/Perl/perl5/blob/blead/INSTALL), which is a bit of a surprise to me. We should fix that. IMO it is much more reasonable to expect these things to be documented in INSTALL than it is to expect them to be documented in pod/

The reality is that we do not have a good place to document these things and we have tension about which repository should own some of this. IMO we should fix both problems. We should just accept that there is a difference between a feature flag/define like `NO_TAINT_SUPPORT` or the defines mentioned in [hv_func.h](https://github.com/Perl/perl5/blob/blead/hv_func.h) and the majority of the settings that are controlled by [./Configure](https://github.com/Perl/perl5/blob/blead/Configure). We should have a specific place to put documentation for settings that are owned by [perl.git](https://github.com/Perl/perl5) as distinct from those that are imported from [metaconfig](https://github.com/Perl/metaconfig), and we should have a standard way to ensure that those settings are introspectable via Config, and we should have a way to be sure that we test that these settings have documentation and are mentioned by [INSTALL](https://github.com/Perl/perl5/blob/blead/INSTALL). I don't know that we should have all settings documented somewhere in pod/, we don't have that now, but it is not unreasonable either. 

Honestly a lot of these issues come up because [./Configure](https://github.com/Perl/perl5/blob/blead/Configure) doesn't live in the same repo as the rest of the perl code. I really really wonder if that makes sense these days where git manages our repo. If we merged the two repos maybe it would go a long way toward solving these issues. Double entry almost guarantees issues as a general rule, and double entry into distinct repos just makes the problem worse.

**Comment by demerphq** [comment]
@demerphq said
> Yeah, I can take care of that.

Fixed in #21047

@demerphq said
> Interestingly it is not mentioned in [INSTALL](https://github.com/Perl/perl5/blob/blead/INSTALL),

Fixed in #21047
